src/Components/Components/src/SupplyParameterFromPersistentComponentStateValueProvider.cs

```diff
@@ -0,0 +1,270 @@
+        }
+        finally
+        {
+            if (pool != null)
+            {
+                ArrayPool<byte>.Shared.Return(pool, clearArray: true);
+            }
+        }
+    }
+
+    private static ReadOnlySpan<char> ResolveKeySpan(object? key)
+    {
+        if (key is IFormattable formattable)
+        {
+            var keyString = formattable.ToString("", CultureInfo.InvariantCulture);
+            return keyString.AsSpan();
+        }
+        else if (key is IConvertible convertible)
+        {
+            var keyString = convertible.ToString(CultureInfo.InvariantCulture);
+            return keyString.AsSpan();
+        }
+        return default;
+    }
+
+    private static void GrowBuffer(ref byte[]? pool, ref Span<byte> keyBuffer, int? size = null)
+    {
+        var newPool = pool == null ? ArrayPool<byte>.Shared.Rent(size ?? 2048) : ArrayPool<byte>.Shared.Rent(pool.Length * 2);
+        keyBuffer.CopyTo(newPool);
+        keyBuffer = newPool;
+        if (pool != null)
+        {
+            ArrayPool<byte>.Shared.Return(pool, clearArray: true);
+        }
+        pool = newPool;
+    }
+
+    private static object? GetSerializableKey(ComponentState componentState)
+    {
+        if (componentState.ParentComponentState is not { } parentComponentState)
+        {
+            return null;
+        }
+
+        // Check if the parentComponentState has a `@key` directive applied to the current component.
+        var frames = parentComponentState.CurrentRenderTree.GetFrames();
+        for (var i = 0; i < frames.Count; i++)
+        {
+            ref var currentFrame = ref frames.Array[i];
+            if (currentFrame.FrameType != RenderTree.RenderTreeFrameType.Component ||
+                !ReferenceEquals(componentState.Component, currentFrame.Component))
+            {
+                // Skip any frame that is not the current component.
+                continue;
+            }
+
+            var componentKey = currentFrame.ComponentKey;
+            return !IsSerializableKey(componentKey) ? null : componentKey;
+        }
+
+        return null;
+    }
+
+    private static string GetComponentType(ComponentState componentState) => componentState.Component.GetType().FullName!;
+
+    private static string GetParentComponentType(ComponentState componentState) =>
+        componentState.ParentComponentState == null ? "" : GetComponentType(componentState.ParentComponentState);
+
+    private static byte[] KeyFactory((string parentComponentType, string componentType, string propertyName) parts) =>
+        SHA256.HashData(Encoding.UTF8.GetBytes(string.Join(".", parts.parentComponentType, parts.componentType, parts.propertyName)));
+
+    private static bool IsSerializableKey(object key)
+    {
+        if (key == null)
+        {
+            return false;
+        }
+        var keyType = key.GetType();
+        var result = Type.GetTypeCode(keyType) != TypeCode.Object
```

Comment:
nit: Should `DateTimeOffset` be supported here as well?

Reply:
Added support for it